README.md

# Milvus Python SDK

[![version](https://img.shields.io/pypi/v/pymilvus-orm.svg?color=blue)](https://pypi.org/project/pymilvus-orm/)
[![Supported Python Versions](https://img.shields.io/pypi/pyversions/pymilvus-orm?logo=python&logoColor=blue)](https://pypi.org/project/pymilvus-orm/)
[![Downloads](https://pepy.tech/badge/pymilvus-orm)](https://pepy.tech/project/pymilvus-orm)
[![Downloads](https://pepy.tech/badge/pymilvus-orm/month)](https://pepy.tech/project/pymilvus-orm/month)
[![Downloads](https://pepy.tech/badge/pymilvus-orm/week)](https://pepy.tech/project/pymilvus-orm/week)
[![license](https://img.shields.io/hexpm/l/plug.svg?color=green)](https://github.com/milvus-io/pymilvus-orm/blob/master/LICENSE)
[![Mergify Status][mergify-status]][mergify]

[mergify]: https://mergify.io
[mergify-status]: https://img.shields.io/endpoint.svg?url=https://gh.mergify.io/badges/milvus-io/pymilvus-orm&style=plastic

Another Python SDK for [Milvus](https://github.com/milvus-io/milvus). To contribute code to this project, please read our [contribution guidelines](https://github.com/milvus-io/milvus/blob/master/CONTRIBUTING.md) first. If you have some ideas or encounter a problem, you can find us in the Slack channel [#py-milvus](https://milvusio.slack.com/archives/C024XTWMT4L).


## Compatibility
The following collection shows Milvus versions and recommended PyMilvus-ORM versions:

|Milvus version| Recommended PyMilvus-ORM version |
|:-----:|:-----:|
| 2.0.0-RC1 | 2.0.0rc1 |


## Installation

You can install PyMilvus-ORM via `pip3` for Python 3.6+:

```shell
# Note this will only install the latest stable version
$ pip3 install pymilvus-orm
```

You can install a specific version of PyMilvus-ORM by:

```shell
$ pip3 install pymilvus-orm==2.0.0rc1
```

You can upgrade PyMilvus-ORM to the latest stable version by:

```shell
$ pip3 install --upgrade pymilvus-orm
```


## Documentation

Documentation is available online: https://pymilvus-orm.readthedocs.io/.


## Packages

### Released packages

The release of PyMilvus ORM is managed on GitHub, and GitHub Actions will package and upload each version to PyPI.

The release version number of PyMilvus ORM follows PEP440, the format is x.y.z, and the corresponding git tag name is vx.y.z (x/y/z are numbers from 0 to 9).

For example, after PyMilvus ORM 1.0.1 is released, a tag named v1.0.1 can be found on GitHub, and a package with version 1.0.1 can be downloaded on PyPI.

### Developing packages (Working in progress)

The commits on the development branch of each version will be packaged and uploaded to Test PyPI. Development branches refer to branches such as 1.0 and 1.1, and version releases are generated from the development branches, such as 1.0.1 and 1.0.2.

The package name generated by the development branch is x.y.z.dev<dist>, where <dist> is the number of commits that differ from the most recent release.

For example, after the release of 1.0.1, two commits were submitted on the 1.0 branch. At this time, the automatic packaging version number of the development branch is 1.0.1.dev2.

To install the package on test.pypi.org, you need to append the parameter --index-url after pip, for example:
```shell
$ python3 -m pip install --index-url https://test.pypi.org/simple/ pymilvus-orm
```


## License
[Apache License 2.0](LICENSE)


CHANGELOG.md
# pymilvus 0.0.1(TBD)

## Task
- Get value of metadata fields from search result output (#90)
- Improve interface documentation (#89)
- Add examples in partition.py (#88)
- add example for index.py (#87)
- Format collection exception raising (#84)
- Add dim param example (#86)
- Add search for hello_milvus (#85)
- Add wrapper to SearchResult-related class (#83)
- Improve hello_milvus file (#81)
- update return format in index.py (#82)
- Move return type behind the return annotation in partition.py (#79)
- update exception in index.py and add index.rst (#80)
- Format returns (#78)
- Improve the docstring of connections.Connections (#77)
- Add interface connections (#76)
- Add annotions in partition.py (#74)
- Add exceptions and errors in collection (#75)
- Add exceptions and errors class (#70)
- Implement and add example to Collection.search (#73)
- Add requirements for installing and add hello_milvus (#72)
- Add index related examples (#71)
- Add util implements (#63)
- Fix partition and add partition example (#67)
- Add interface documentation about collection (#69)
- Update unittest (#68)
- Test trigger readocs (#66)
- Add case when CollectionSchema specifies the primary field (#61)
- Referencing other interfaces in interface comments (#65)
- Fix todo lists (#64)
- Update todo items (#62)
- Fix mistake return type (#60)
- Add TODO marks (#59)
- Fix bug for inserting when collection schema has a primary field (#58)
- Add create index test for binary data (#56)
- Update tests in test_partition (#57)
- Add codecov config file (#55)
- Add check in some interfaces (#54)
- Add test for create index (#53)
- Update example (#52)
- Add copyright and modify params and documentation (#51)
- Delete auto_id param for collection schema (#49)
- Add an example (#50)
- Add example for testing collection (#48)
- Add codecov badge in README.md (#47)
- Update GitHub Actions (#46)
- Fix some partition unittests (#45)
- Add partition impl in MockMilvus (#44)
- Update mergify (#43)
- Implement mock milvus (#42)
- Update index and partition unittests (#41)
- Implement Collection.insert, Support list and tuple (#40)
- Add example for collection init method (#36)
- Update doc: add util doc
- Update github actions (#35)
- Update unittest (#32)
- Add to_dict for CollectionSchema and FieldSchema (#31)
- Implement the init method of the v1 of collection (#26)
- Fix bug for html bug (#29)
- Fix bug: access attribute from None (#25)
- Add example to Collection.load & Collection.release (#28)
- Update index interfaces (#27)
- Impl CollectionSchema and FieldSchema (#24)
- Add example for proprety methods of collection(#17)
- Mock milvus.Milvus in test_connections.py (#21)
- Fix compile error (#23)
- Add mergify file (#22)
- Update interface for partition module (#20)
- Update documentation (#19)
- Realized the properties of collection (#16)
- Update interfaces in collection module (#14)
- Update interfaces in utility module (#11)
- Implement the management class of milvus connections (#10)
- Add GitHub Actions to run unittests (#12)
- rename milvus_orm to pymilvus_orm (#9)
- Update doc (#8)
- Add the annotation of apis (#7)
- Add annotation for apis (#6)
- Add sdk interface draft (#5)
- Initialize the api framework #2
- Initialize the api framework
- Establish the initial directory structure and copy the document files (#1)
- Initial commit

examples/partition.py
# Copyright (C) 2019-2020 Zilliz. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
# with the License. You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License
# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
# or implied. See the License for the specific language governing permissions and limitations under the License.

from pymilvus_orm.collection import Collection
from pymilvus_orm.connections import connections
from pymilvus_orm.partition import Partition
from pymilvus_orm.schema import FieldSchema, CollectionSchema
from pymilvus_orm.types import DataType
from pymilvus_orm.utility import list_collections, has_partition
import random
from sklearn import preprocessing
import string

default_dim = 128
default_nb = 3000
default_float_vec_field_name = "float_vector"
default_segment_row_limit = 1000



all_index_types = [
    "FLAT",
    "IVF_FLAT",
    "IVF_SQ8",
    # "IVF_SQ8_HYBRID",
    "IVF_PQ",
    "HNSW",
    # "NSG",
    "ANNOY",
    "RHNSW_FLAT",
    "RHNSW_PQ",
    "RHNSW_SQ",
    "BIN_FLAT",
    "BIN_IVF_FLAT"
]

default_index_params = [
    {"nlist": 128},
    {"nlist": 128},
    {"nlist": 128},
    # {"nlist": 128},
    {"nlist": 128, "m": 16, "nbits": 8},
    {"M": 48, "efConstruction": 500},
    # {"search_length": 50, "out_degree": 40, "candidate_pool_size": 100, "knng": 50},
    {"n_trees": 50},
    {"M": 48, "efConstruction": 500},
    {"M": 48, "efConstruction": 500, "PQM": 64},
    {"M": 48, "efConstruction": 500},
    {"nlist": 128},
    {"nlist": 128}
]


default_index = {"index_type": "IVF_FLAT", "params": {"nlist": 128}, "metric_type": "L2"}


def gen_default_fields(auto_id=True):
    default_fields = [
        FieldSchema(name="count", dtype=DataType.INT64, is_primary=True),
        FieldSchema(name="float", dtype=DataType.FLOAT),
        FieldSchema(name=default_float_vec_field_name, dtype=DataType.FLOAT_VECTOR, dim=default_dim)
    ]
    default_schema = CollectionSchema(fields=default_fields, description="test collection",
                                      segment_row_limit=default_segment_row_limit, auto_id=False)
    return default_schema


def gen_vectors(num, dim, is_normal=True):
    vectors = [[random.random() for _ in range(dim)] for _ in range(num)]
    vectors = preprocessing.normalize(vectors, axis=1, norm='l2')
    return vectors.tolist()


def gen_data(nb, is_normal=False):
    vectors = gen_vectors(nb, default_dim, is_normal)
    entities = [
        [i for i in range(nb)],
        [float(i) for i in range(nb)],
        vectors
    ]
    return entities


def gen_unique_str(str_value=None):
    prefix = "".join(random.choice(string.ascii_letters + string.digits) for _ in range(8))
    return "collection_" + prefix if str_value is None else str_value + "_" + prefix


def binary_support():
    return ["BIN_FLAT", "BIN_IVF_FLAT"]


def gen_simple_index():
    index_params = []
    for i in range(len(all_index_types)):
        if all_index_types[i] in binary_support():
            continue
        dic = {"index_type": all_index_types[i], "metric_type": "L2"}
        dic.update({"params": default_index_params[i]})
        index_params.append(dic)
    return index_params

def test_partition():
    connections.connect(alias="default")
    print("create collection")
    collection = Collection(name=gen_unique_str(), schema=gen_default_fields())
    print("create partition")
    partition = Partition(collection, name=gen_unique_str())
    print(list_collections())
    assert has_partition(collection.name, partition.name) == True

    data = gen_data(default_nb)
    print("insert data to partition")
    partition.insert(data)
    assert partition.is_empty is False
    assert partition.num_entities == default_nb

    print("load partition")
    partition.load()
    topK = 5
    search_params = {"metric_type": "L2", "params": {"nprobe": 10}}
    print("search partition")
    res = partition.search(data[2][-2:], "float_vector", search_params, topK, "count > 100")
    for hits in res:
        for hit in hits:
            print(hit)

    print("release partition")
    partition.release()
    print("drop partition")
    partition.drop()
    print("drop collection")
    collection.drop()

test_partition()
